### ISSUE TITLE
Kernel soft lockup bug while manually flushing two CAS instances in parallel

### ISSUE THREAD

**Issue by eriksun421** [Open-CAS/open-cas-linux#400]
OS: CentOS 7.4 (3.10.0-693.el7.x86_64)
CAS: master branch - HEAD commit 68533834

Reproduce steps:
1. Use Intel P4800X to create two CAS instances, each one attached 6 HDD drives as core devices
2. CAS is running in 'wb' mode with 32K cache line size and 'nop' cleaning policy
3. Generate write workloads to make both caching devices 100% dirty
4. manually flush two CAS instances in parallel by "casadm -F -i X"
5. Repeat steps 3&4 until kernel soft-lockup bug occurs

Symptom logs:
[3871402.444349] cache2: Flushing cache
[3871403.808361] cache2: Flushing cache completed
[3871407.225162] cache1: Flushing cache
[3871408.613754] cache1: Flushing cache completed
[3978987.226059] cache1: Flushing cache
[3979008.700426] cache2: Flushing cache
[3979016.917189] NMI watchdog: BUG: soft lockup - CPU#24 stuck for 23s! [cas_io_cache1_-:5224]
[3979016.926092] Modules linked in: sctp_diag sctp dccp_diag dccp unix_diag binfmt_misc cas_cache(OE) cas_disk(OE) tcp_diag udp_diag inet_diag bonding skx_edac edac_core coretemp intel_rapl iosf_mbi kvm_intel kvm irqbypass crc32_pclmul ghash_clmulni_intel aesni_intel lrw iTCO_wdt gf128mul glue_helper ablk_helper iTCO_vendor_support ses enclosure cryptd sg joydev nvme pcspkr nvme_core lpc_ich i2c_i801 ipmi_si shpchp ipmi_devintf mei_me ipmi_msghandler mei wmi nfit libnvdimm acpi_power_meter nfsd auth_rpcgss nfs_acl lockd grace sunrpc ip_tables xfs libcrc32c sd_mod crc_t10dif crct10dif_generic ast i2c_algo_bit drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops ttm i40e ixgbe mpt3sas drm ahci crct10dif_pclmul crct10dif_common libahci mdio raid_class crc32c_intel ptp scsi_transport_sas libata dca
[3979017.001792]  i2c_core pps_core dm_mirror dm_region_hash dm_log dm_mod
[3979017.007629] CPU: 24 PID: 5224 Comm: cas_io_cache1_- Tainted: G          IOEL ------------   3.10.0-693.el7.x86_64 #1
[3979017.018877] Hardware name: Inspur SA5212M5/YZMB-00882-102, BIOS 4.0.3 03/29/2018
[3979017.027004] task: ffff882e0f489fa0 ti: ffff882a3165c000 task.ti: ffff882a3165c000
[3979017.035210] RIP: 0010:[<ffffffffc05fc6f7>]  [<ffffffffc05fc6f7>] _ocf_cleaner_cmp+0x27/0x80 [cas_cache]
[3979017.045361] RSP: 0018:ffff882a3165fc90  EFLAGS: 00000282
[3979017.051413] RAX: 0000000000000006 RBX: ffff8800054e4f10 RCX: 0000000000000006
[3979017.059291] RDX: 000000000092ad80 RSI: ffffc9022ffdad80 RDI: ffffc9022ffdad70
[3979017.067175] RBP: ffff882a3165fca0 R08: 00000000004956b0 R09: 00000000001fcc70
[3979017.075060] R10: 0000000000000004 R11: 0000000000000005 R12: ffff882a3165fc18
[3979017.082953] R13: ffff881ffe916cc0 R14: 0000000000000018 R15: ffff881ffe916cc0
[3979017.090850] FS:  0000000000000000(0000) GS:ffff881ffe900000(0000) knlGS:0000000000000000
[3979017.099712] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[3979017.106240] CR2: 00007f68de6d94f0 CR3: 0000002c8dbfa000 CR4: 00000000003407e0
[3979017.114168] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[3979017.122098] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[3979017.130017] Stack:
[3979017.132835]  ffff882a3165fca8 0000000000000018 ffff882a3165fd10 ffffffffc05cd72d
[3979017.141116]  ffffffffffffffe0 fffffffffffffff0 ffffffff00000010 0000000004e5bc80
[3979017.149408]  0000000004e5bc70 00000000004956b0 000000000092ad80 0000000000000006
[3979017.157709] Call Trace:
[3979017.160998]  [<ffffffffc05cd72d>] env_sort+0x1ad/0x210 [cas_cache]
[3979017.168028]  [<ffffffffc05fd993>] ocf_cleaner_sort_flush_containers+0x53/0x60 [cas_cache]
[3979017.177054]  [<ffffffffc05f1900>] ? _ocf_mngt_flush_core_complete+0x80/0x80 [cas_cache]
[3979017.185905]  [<ffffffffc05f1dbd>] _ocf_mngt_flush_containers+0x3d/0x1a0 [cas_cache]
[3979017.194429]  [<ffffffffc05f24c2>] _ocf_mngt_cache_flush+0x92/0x500 [cas_cache]
[3979017.202523]  [<ffffffffc05ff181>] _ocf_pipeline_run_step+0x61/0xd0 [cas_cache]
[3979017.210616]  [<ffffffffc05f72a3>] ocf_io_handle+0x23/0x40 [cas_cache]
[3979017.217930]  [<ffffffffc05f7305>] ocf_queue_run_single+0x45/0x50 [cas_cache]
[3979017.225850]  [<ffffffffc05f7338>] ocf_queue_run+0x28/0x50 [cas_cache]
[3979017.233174]  [<ffffffffc05cf45a>] _cas_io_queue_thread+0xfa/0x170 [cas_cache]
[3979017.241178]  [<ffffffff810b1910>] ? wake_up_atomic_t+0x30/0x30
[3979017.247890]  [<ffffffffc05cf360>] ? cas_service_ioctl_ctrl+0x1bd0/0x1bd0 [cas_cache]
[3979017.256497]  [<ffffffff810b098f>] kthread+0xcf/0xe0
[3979017.262248]  [<ffffffff810b08c0>] ? insert_kthread_work+0x40/0x40
[3979017.269218]  [<ffffffff816b4f18>] ret_from_fork+0x58/0x90
[3979017.275494]  [<ffffffff810b08c0>] ? insert_kthread_work+0x40/0x40
[3979017.282460] Code: 00 00 00 00 0f 1f 44 00 00 55 48 89 e5 48 83 ec 10 8b 05 21 10 01 00 83 c0 01 3d 40 42 0f 00 89 05 13 10 01 00 74 2d 0f b7 47 0c <0f> b7 56 0c 66 39 d0 74 10 66 39 c2 c9 19 c0 83 e0 02 83 e8 01 
[3982234.546598] cache1: Flushing cache completed
[3982338.776354] cache2: Flushing cache completed
[3983423.308927] cache2: Flushing cache
[3983423.313308] cache2: Flushing cache completed

**Comment by imjfckm** [comment]
Hi @eriksun421, thanks for your report.

What you're observing is caused by CAS sorting the drity cache lines to flush and not yielding to the scheduler while doing it. If you have enough dirty cache lines to sort this process duration may exceed soft lockup threshold.

It is not dangerous in any way either to data or availability, but it's not desirable ;)

I'll create a new PR with fix for this issue today.

**Comment by eriksun421** [comment]
@imjfckm, Glad to hear your response quickly. Thanks for root causing it and fixing it in such a short time.
